x-pack/plugins/event_log/server/event_logger.ts

```diff
@@ -0,0 +1,179 @@
+import { Logger } from 'src/core/server';
+import { merge } from 'lodash';
+
+import { Plugin } from './plugin';
+import { EsContext } from './es';
+import {
+  IEvent,
+  IValidatedEvent,
+  IEventLogger,
+  IEventLogService,
+  ECS_VERSION,
+  EventSchema,
+} from './types';
+
+type SystemLogger = Plugin['systemLogger'];
+
+interface Doc {
+  index: string;
+  body: IEvent;
+}
+
+interface IEventLoggerCtorParams {
+  esContext: EsContext;
+  eventLogService: IEventLogService;
+  initialProperties: IEvent;
+  systemLogger: SystemLogger;
+}
+
+export class EventLogger implements IEventLogger {
+  private esContext: EsContext;
+  private eventLogService: IEventLogService;
+  private initialProperties: IEvent;
+  private systemLogger: SystemLogger;
+
+  constructor(ctorParams: IEventLoggerCtorParams) {
+    this.esContext = ctorParams.esContext;
+    this.eventLogService = ctorParams.eventLogService;
+    this.initialProperties = ctorParams.initialProperties;
+    this.systemLogger = ctorParams.systemLogger;
+  }
+
+  startTiming(event: IEvent): void {
+    if (event == null) return;
+    if (event.event == null) event.event = {};
+
+    event.event.start = new Date().toISOString();
+  }
+
+  stopTiming(event: IEvent): void {
+    if (event == null) return;
+    if (event.event == null) return;
+
+    const start = getEventStart(event);
+    if (start == null || isNaN(start)) return;
+
+    const end = Date.now();
+    event.event.end = new Date(end).toISOString();
+    event.event.duration = (end - start) * 1000 * 1000; // nanoseconds
+  }
+
+  // non-blocking, but spawns an async task to do the work
+  logEvent(eventProperties: IEvent): void {
+    if (!this.eventLogService.isEnabled()) return;
+
+    const event: IEvent = {};
+
+    // merge the initial properties and event properties
+    merge(event, this.initialProperties, eventProperties);
+
+    // add fixed properties
+    event['@timestamp'] = new Date().toISOString();
+    if (event.ecs == null) event.ecs = {};
+    event.ecs.version = ECS_VERSION;
+
+    // TODO add kibana server uuid
+    // if (event.kibana == null) event.kibana = {};
+    // event.kibana.server_uuid = NP version of config.get('server.uuid');
+
+    let validatedEvent: IValidatedEvent;
+    try {
+      validatedEvent = validateEvent(this.eventLogService, event);
+    } catch (err) {
+      this.systemLogger.warn(`invalid event logged: ${err.message}`);
+      return;
+    }
+
+    const doc: Doc = {
+      index: this.esContext.esNames.alias,
+      body: validatedEvent,
+    };
+
+    if (this.eventLogService.isIndexingEntries()) {
+      indexEventDoc(this.esContext, doc);
+    }
+
+    if (this.eventLogService.isLoggingEntries()) {
+      logEventDoc(this.systemLogger, doc);
+    }
+  }
+}
+
+// return the epoch millis of the start date, or null; may be NaN if garbage
+function getEventStart(event: IEvent): number | null {
+  if (event == null || event.event == null || event.event.start == null) return null;
```

Comment:
do we?  :-) (in reference to recent CI failure complaining about optional chaining).  I'm anxious to start using this, so will change if we can!

Reply:
So, we have it in TS, but seems JS code breaks on this even though the Node version we use now supports it (I was updating a reference in another plugin which is still JS).
I think it's because we're using a transpiler that targets _lower_ than the version of Node we use....I need to look into it, but in anycase, I've been using it in TS code without a hitch.